compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -113,7 +121,218 @@ static void specializeGenericTransposeOp(RewriterBase &rewriter,
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp) {
+    return failure();
+  }
+  // TODO: Right now this is restricted to contractions due to fragility around
+  // handling of convolutions.
+  if (!isa<linalg::GenericOp>(linalgOp) &&
+      !(allowGeneralizing && linalg::isaContractionOpInterface(linalgOp))) {
+    return failure();
+  }
+  auto genericOp = dyn_cast<linalg::GenericOp>(op);
+  if (genericOp) {
+    return genericOp;
+  }
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(linalgOp);
+  return linalg::generalizeNamedOp(rewriter, linalgOp);
+}
+
+//===----------------------------------------------------------------------===//
+// Transpose Bubbling Patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+// Fuses a transpose with the init of a linalg.generic op or contraction op.
+// Contraction ops are generalized and then treated as a generic. For example,
+//
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>) outs(%1 : tensor<7x2x5>)
+//
+// %2 = linalg.transpose ... permutation = [0, 2, 1] :
+//                           tensor<7x2x5> -> tensor<7x5x2>
+// Becomes
+//
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0, d1)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>) outs(%3 : tensor<7x5x2>)
+class BubbleTransposeThroughDpsInit
+    : public OpRewritePattern<linalg::TransposeOp> {
+public:
+  using OpRewritePattern<linalg::TransposeOp>::OpRewritePattern;
+  BubbleTransposeThroughDpsInit(MLIRContext *ctx, bool aggressiveProp,
+                                PatternBenefit b = 1)
+      : OpRewritePattern<linalg::TransposeOp>(ctx, b),
+        allowGeneralizing(aggressiveProp) {}
+
+  LogicalResult matchAndRewrite(linalg::TransposeOp transposeOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(transposeOp)) {
+      return failure();
+    }
+    OpResult result = dyn_cast<OpResult>(transposeOp.getInput());
+    if (!result) {
+      return failure();
+    }
+    if (!result.hasOneUse()) {
+      return failure();
+    }
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(result.getOwner());
+    if (!linalgOp) {
+      return failure();
+    }
+
+    int64_t resultIndex = result.getResultNumber();
+    auto maybeGenericOp = getGenericOpOrGeneralizeContraction(
+        rewriter, result.getOwner(), allowGeneralizing);
+    if (failed(maybeGenericOp)) {
+      return failure();
+    }
+    auto genericOp = maybeGenericOp.value();
+    result = genericOp->getOpResult(resultIndex);
+
+    ArrayRef<int64_t> perm = transposeOp.getPermutation();
+    auto invPerm = invertPermutationVector(perm);
+
+    OpOperand *init = genericOp.getDpsInitOperand(resultIndex);
```

Comment:
I think you can just use `getDpsInit(resultIndex)` and avoid getting the `OpOperand`